qiskit/algorithms/optimizers/suffix_averaging.py

```diff
@@ -0,0 +1,154 @@
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Suffix Averaging optimier"""
+
+import warnings
+from typing import Callable, List, Optional, Tuple
+
+import numpy as np
+
+from qiskit.algorithms.optimizers import NFT, QNSPSA, SPSA, GradientDescent
+
+from .optimizer import POINT, Optimizer, OptimizerResult
+
+
+class SuffixAveragingOptimizer(Optimizer):
+    r"""The suffix averaging optimizer.
+
+    Given a sequence of circuit parameters of parameterized quantum circuit
+    obtained from :math:`T` iterations, :math:`\{\vec{\theta}^{(t)}\}_{t=1}^{t=T}`,
+    the SuffixAveragingOptimizer returns the averaged point of the last n_params_suffix
+    points,
+
+    .. math::
+
+        \overline{\vec{\theta}} = \frac{1}{n_params_suffix} \sum_{t=T-n_params_suffix-1}^{T}
+        \vec{\theta}^{(t)}.
+
+    Examples:
+        .. code-block::python
+            from qiskit import Aer
+            from qiskit.utils import QuantumInstance
+            from qiskit.algorithms import VQE
+            from qiskit.opflow.gradients import Gradient
+            from qiskit.circuit.library import TwoLocal
+            from qiskit.algorithms.optimizers import ADAM
+            from qiskit.algorithms.optimizers.suffix_averaging import SuffixAveragingOptimizer
+
+            backend = Aer.get_backend("qasm_simulator")
+            quantum_instance = QuantumInstance(backend=backend, shots=1000)
+            ansatz = TwoLocal(4, ["rx", "rz"], "cx", "linear")
+            optimizer = ADAM(maxiter=1000)
+            suffix_optimizer = SuffixAveragingOptimizer(optimizer, n_params_suffix=50)
+            gradient = Gradient(grad_method="param_shift")
+            vqe = VQE(
+                ansatz = ansatz,
+                optimizer = suffix_optimizer,
+                gradient = gradient,
+                quantum_instance = quantum_instance
+            )
+
+    References:
+        [1] S. Tamiya and H. Yamasaki. 2021.
+        Stochastic Gradient Line Bayesian Optimization
+        for Efficient Noise-Robust Optimization of Parameterized Quantum Circuits.
+        arXiv preprint arXiv:2111.07952.
+    """
+
+    def __init__(self, optimizer: Optimizer, n_params_suffix: int = 50) -> None:
+        """
+        Args:
+            optimizer: The optimizer used for optimizing parameterized quantum circuits.
+            n_params_suffix: The number of circuit parameters for taking the suffix average.
+        """
+
+        self._n_params_suffix = n_params_suffix
+        self._optimizer = optimizer
+
+        self._circ_params = []
+
+        original_callback = self._optimizer.callback
+        if isinstance(self._optimizer, (SPSA, QNSPSA)):
```

Comment:
So the `callback` signature added, that is compatible with scipy, is different than that supplied by (QN)SPSA and GradientDescent. I see SPSA has the same public instance var (I guess that was changed not so long back and does not have any of its many public instance vars documented which you can see [here ](https://qiskit.org/documentation/stubs/qiskit.algorithms.optimizers.SPSA.html#qiskit.algorithms.optimizers.SPSA)they only appear on constructor - this is why made the comment on the callback in the base class).

It would be nice to find someway to deal with this difference in callback signatures from a user perspective. Clearly these classes do not conform to an OptimizerCallback when their callback set read/set via the callback instance var,